throw launch

zeroos opened the pull request:
This PR implements a throw-launch feature for multicopters: the ability to start the drone by throwing it into the air. The drone should not spin the propellers while held in hand and only do that when airborne.

It is based on the discussions here: https://discuss.px4.io/t/quadrotor-launch-control-throw-launches/212

Technically, the feature is realized by locking the motors (i.e., preventing them from spinning) after the drone is armed and releasing the lock only when a fall down is detected. Then, the drone stabilizes according to the mode it is in.

The feature was tested both in simulation and real flight, an example flight log is available here: 
https://logs.px4.io/plot_app?log=184b1c7f-e875-4937-a12d-c7dfca8e70aa (bad throw with lots of rolling, but the drone still recovers)
https://logs.px4.io/plot_app?log=c15d061c-538f-4134-85f8-83058882073e

The documentation is available here:
https://github.com/zeroos/PX4-user_guide/blob/f31812a7ac233697c24bd8223b7b3c7ba4914ab0/en/advanced_features/throw_launch.md?plain=1#L23

### Changelog 

#### 20 Feb 2023
A first working version was tested both in simulation and reality.

#### 24 Feb 2023 
The code was refined and more tests were conducted, sometimes the propellers slightly spin while arming. It's probably due to the fact that the locking of motors happens after arming and there might be a slight delay. Would be nice to modify the solution to take care of this.

#### 18 Apr 2023

I was slightly delayed due to some platform changes, but on the bright side, the feature was tested using two different drones (albeit very similar, as both are based on an X250 racer platform, one with Pixhawk 6 and the other one with Pixhawk 4).

Another factor that delayed the work was me trying to make "hold" mode work during the throw and change the setpoint while in flight. In the end, I concluded it's better to throw the drone in the "position" mode as it behaves more predictably even when the position estimate is imperfect due to the throw.

There were also some changes done during this iteration, they revolved mainly around suppressing automatic disarming during the throw. I also needed to allow the position controller to skip the "takeoff" phase when freefall is detected after arming.

Added flight log from a successful flight.

#### 24 Apr 2023

I've added the documentation and tested everything once more. Looking forward for feedback!

Squashed all commits; the full history is here: https://github.com/zeroos/PX4-Autopilot/tree/throw-mode-history

#### 25 Apr 2023

I've realized some automated tests were not passing. Now, they should be fine.

I also rebased the branch to be on top of the current main branch and performed another flight test to make sure everything is still good. The flight log is attached to this message.

#### 15 Aug 2023

Rebased the code on top of the most recent PX4 version and provided new logs. The old logs are here: 
https://logs.px4.io/plot_app?log=5583e2c8-ec95-41aa-abdd-873a09df5477 (before rebase to main, 18th of April),
https://logs.px4.io/plot_app?log=484e9ee6-751d-4bd4-a123-6b8d33d46652 (after rebase on the 25th of April).

### TODO:

- [x] add parameter to enable/disable the feature
- [x] replace throw_mode to throw_launch across the code
- [x] come up with a way to keep the "auto disarm" feature still working
- [x] replace debug prints with some useful information
- [x] consider adding some beeps when the drone is armed, but motors are locked
- [x] try to find someone experienced who can confirm this way of locking motors (and unlocking) is okay
- [x] propose a safety mechanism to make sure the drone does not start if accidentally dropped
- [x] modify the code to prevent the slight movement of the motors after arming
- [x] provide documentation
- [x] modify the PR to work with the main version of PX4
- [x] provide new flight logs

swimmingseeds commented:
This pull request has been mentioned on **Discussion Forum for PX4, Pixhawk, QGroundControl, MAVSDK, MAVLink**. There might be relevant details there:

https://discuss.px4.io/t/quadrotor-launch-control-throw-launches/212/27

bresch commented:
Don't you think that solely relying on the freefall detection to tell that the copter was thrown might lead to catastrophic false positives if the user initially quickly lowers the altitude of the drone before throwing it in the air? 
Maybe an hysteresis is enough to increase safety? Maybe requiring an initial upward or horizontal velocity would prevent triggering if the drone is just dropped.

zeroos commented:
Thanks @bresch, good suggestion! In practice, it is not easy to trigger the fall-down detection by accident on the drone I'm using (a typical X250 racer frame with Pixhawk 6X), but I can imagine some other builds might be more sensitive or someone might drop a  drone by accident.

I was thinking about how to realize it while still maintaining functionality (maybe someone wants to drop a drone or throw it down the cliff). I see three options: 
(1) having the parameter COM_THROW_EN, which is an ENUM, where 0 means the throw-launch feature is disabled, 1 means it's enabled and requires initial hard-coded speed (e.g., 1 m/s horizontal and vertical -- to be tested), and 2 means the feature is enabled without any safety checks (like it's implemented now) and can be used for drops
(2) adding parameters COM_THROW_MIN_V_VERT and COM_THROW_MIN_V_HOR specifying minimum vertical and horizontal speed for the feature to activate with some reasonable defaults
(3) just add this safety feature with reasonable defaults and always enforce it

Please let me know what you think!

As a side note, I wanted to attend the dev-call yesterday to ask if this contribution is even welcome and if the approach feels reasonable, but unfortunately I've had some errands to run and did not manage on time. I'm especially worried about the use of `_actuator_armed.lockdown` to lock (and later automatically unlock) the motors. Are you able to give some feedback on this or should I wait for the next dev call?

swimmingseeds commented:
This pull request has been mentioned on **Discussion Forum for PX4, Pixhawk, QGroundControl, MAVSDK, MAVLink**. There might be relevant details there:

https://discuss.px4.io/t/px4-community-q-a-march-01-2023/30941/3

swimmingseeds commented:
This pull request has been mentioned on **Discussion Forum for PX4, Pixhawk, QGroundControl, MAVSDK, MAVLink**. There might be relevant details there:

https://discuss.px4.io/t/px4-community-q-a-march-01-2023/30941/1

zeroos commented:
@bresch it feels like the test that was not going through was based on some assumption that I don't fully understand. The modified test is here: https://github.com/PX4/PX4-Autopilot/pull/21170/files#diff-f7f07514f6a6af2591922c929b5388a0233dd616360630d8fe94da75bde67a9fR57 

Before my changes, the test was written like the `landed` argument could change to false even before `want_takeoff` changed to true. In the tests I did, that does not happen, the drone properly detects that it is landed until the user actually pushes the throttle up and the drone starts. I modified the test to closer represent that.

zeroos commented:
Hello guys! Please let me know if there's anything I can do to push this PR forward. I'd be happy to eventually see this merged.

zeroos commented:
Hello everyone! We did some extensive testing of this feature in the past months and found some possible improvements (e.g., freefall detection does not work well with heavy wind). I could try to rebase it on top of the current master and include our improvements, but I'm not sure if there is interest in this feature as implemented for this PR?

bresch commented:
Hi, good to hear that you made progress on this. I'm pretty sure that several users are (silently) interested in this. Le's make sure it's nicely implemented and I don't see why we shouldn't add this feature to PX4.

zeroos commented:
Hello @bresch , nice to hear from you!

I went through the code of our project and the only thing that was change since the previous PR and is related to the throw launch is the fact that we stopped using freefall as one of the triggers. During one day of our experiments, there was a heavy wind. In these conditions, the freefall never triggered, so we decided not to use it as one of the checks to start spinning the motors. Instead, we check if the vertical speed is negative, which worked for us reliably.

Unfortunately, I see that with the newest master, some checks don't pass. How should I handle this? All of the checks were passing in the previous version (from the end of April).

Please let me know if you have some suggestions on how to improve the code.

bresch commented:
Instead of merging main into your branch I think it would be easier to rebase your branch on top of main. Some submodules are not up-to-date in your branch, that's why CI isn't passing.

zeroos commented:
Thanks for the hint, that was indeed the problem!

In the meantime, some things changed in commander.cpp and additional changes are needed. I'll work on it after the weekend and also I'll try to have some logs from flights with the newest version.

zeroos commented:
Hello @bresch!

I've applied the changes needed for the code to work after the rebase (the way the arming status was checked changed). You will find the new logs here: 

https://logs.px4.io/plot_app?log=184b1c7f-e875-4937-a12d-c7dfca8e70aa (bad throw with a high rotation, but the drone still recovers)
https://logs.px4.io/plot_app?log=c15d061c-538f-4134-85f8-83058882073e

The tests seem to be successful, except of two, where the same test case fails. I believe it's not related to this PR, but please correct me if I'm wrong.

I hope we can get it through this time! :) Looking forward to your feedback.

zeroos commented:
Hey @bresch!

Do you have any feedback about the current state of this PR? I do have some time this week, so if there's anything that needs to be done I could immediately jump on it.

Best regards,
Michał

bresch commented on src/modules/commander/Commander.cpp:
```diff
@@ -1984,6 +1987,47 @@ void Commander::safetyButtonUpdate()
+		float min_launch_speed = _param_com_throw_min_speed.get();
+
+		if (_throw_launch_state == ThrowLaunchState::ARMED &&
+		    vehicle_speed_squared >= min_launch_speed * min_launch_speed) {
```
you can use `Vector3f(_vehicle_local_position.vx, _vehicle_local_position.vy, _vehicle_local_position.vz).longerThan(_param_com_throw_min_speed.get())`

`longerThan`also makes the comparison using the norm squared to avoid a sqrt

zeroos replied:
Nice suggestion, together with the switch case looks better now!

bresch commented on src/modules/commander/Commander.cpp:
```diff
@@ -2123,7 +2173,8 @@ void Commander::handleAutoDisarm()
 				disarm(arm_disarm_reason_t::kill_switch, true);
 
-			} else {
+			} else if (!_param_com_throw_en.get()) {  // don't disarm if throw
```
maybe rather do the skip where it's also done for HITL a couple of lines above

zeroos replied:
Good catch! It also motivated me to add a helper function `isThrowLaunchActive` to do the check instead of checking for appropriate states each time. I've pushed it with the other suggested changes.

bresch commented on src/modules/mc_pos_control/Takeoff/Takeoff.cpp:
```diff
@@ -74,6 +75,9 @@ void TakeoffHandling::updateTakeoffState(const bool armed, const bool landed, co
 			_takeoff_state = TakeoffState::rampup;
 			_takeoff_ramp_progress = 0.f;
 
+		} else if (!landed) {
```
@MaEtUgR Is that fine?

zeroos replied:
@MaEtUgR , to give you some context here: before the change, the position controller to start working required the "want_takeoff" to be true, which means that a pilot needed to put the throttle up. It resulted in a behavior that I found counterintuitive when we put the drone directly into the air after arming: in other modes we tested (hold, stabilized, offboard) the drone behaved as expected (i.e., just started flying), but in position, if we forgot to raise the throttle up, the drone would just fall down, slightly spinning the propellers.

This change does not require the "want_takeoff" to happen if the drone is thrown in the air after arming.

MaEtUgR replied:
But that means if you run any drone and kick it on the ground it will start to correct the position in place and tip over.

MaEtUgR replied:
The case where you arm in other modes and later on switch to a climb rate enabled mode is handled with this: https://github.com/PX4/PX4-Autopilot/blob/23cf0a7317805528d8cdb57806efff83b87895f0/src/modules/mc_pos_control/MulticopterPositionControl.cpp#L564-L566

Can't the case where you throw the drone not be handled explicitly in that way? E.g. when you're in throw mode you are idling ready in the "about to be thrown" state and jump to enabling all controllers skipping the takeoff ramp in an explicit place.

Compared to the suggested whenever a multicopter is idling on the ground and moves a bit too much it jumps up.

zeroos replied:
My main motivation to choose what is proposed was that for me the fact that the drone does not start regulating position after finding itself in the air was counterintuitive (we actually crashed twice before we realized what's happening). We've never had problems with drones starting flying unexpectedly with this firmware, but we treat all armed drones with care, so it never happened that a drone was moved/kicked after being armed.

Anyway, I do understand that we might not want to change the existing behavior to implement the throw launch feature. However, as throw launch is not implemented as a separate mode, the line you pointed out does not automatically trigger, so some modification is required. Do you maybe have some suggestions on how to do that without introducing any mode switching?

The option I was considering was to modify the MulticopterPositionControl and add an if statement checking if THROW_LANUCH parameter is enabled. If so, explicitly skip the takeoff. Is that something you find acceptable?

bresch commented on src/modules/commander/Commander.cpp:
```diff
@@ -1984,6 +1987,47 @@ void Commander::safetyButtonUpdate()
+			_throw_launch_state = ThrowLaunchState::IDLE;
+		}
+
+		if (_throw_launch_state == ThrowLaunchState::IDLE && isArmed()) {
```
A bit of nitpicking here, but a `switch` structure could highlight the state machine a bit better:
```
switch (_throw_launch_state) {
case ThrowLaunchState::IDLE:
	if (isArmed()) {
		mavlink_log_info(&_mavlink_log_pub, "The vehicle is ARMED with throw launch enabled. Throw the vehicle now.\t");
		_throw_launch_state = ThrowLaunchState::ARMED;
	}
	break;
case ThrowLaunchState::ARMED:
....
```

zeroos replied:
Good point, indeed looks better now!

zeroos commented:
Thank you for the code review! Nice suggestions, I've applied all of them and also refined slightly the way we are checking if the throw mode is active or not, I hope you like the changes. 

I've tested it both in simulation and with one of our PX4 Vision drones, the logs are here: 
https://logs.px4.io/plot_app?log=44346e46-8ace-46f9-9c9e-7c9df5bb0307

MaEtUgR commented:
@zeroos Thanks for your contribution and the patience! It was nice talking to you at the dev summit and I got a good idea of the feature :+1: I picked it up and first just rebased on the latest main branch to resolve a small conflict with parameter order. I'd move all the state and functionality into a class now.

MaEtUgR commented:
I added a commit that moves throw-mode exclusive things into a separate class. I also switched to using events for user feedback. The takeoff state machine change is still todo. We should rather skip the ramp-up for this particular case.

zeroos commented:
wow, nice, thanks! It was indeed a pleasure to meet you and talk during the summit. I took a week off after it to explore a bit the US. I will be back in the office after Wednesday next week, I'll try to complete the missing parts then. Thanks!

zeroos commented:
I have reverted the changes to the Takeoff done previously and added the takeoff skip if throw launch is enabled.

Unfortunately, I did not have time this week to test with hardware, but it works in the simulation. Next week (hopefully around Tuesday/Wednesday), I'll try to test with hardware.

zeroos commented:
@MaEtUgR I managed to do the test today and the throw was successful, throwing seems to work fine. You will find the logs here: https://logs.px4.io/plot_app?log=77b156d1-ee00-4d99-be44-a84066a3295f

Unfortunately, the behaviour on my drone on the current main branch is a bit strange, I've described it here: https://github.com/PX4/PX4-Autopilot/issues/22318 I'm sure it also appears on the main branch (not only this PR) and I was still able to fly with it, but the logs have the "Pilot took over using sticks" messages because of it and when the drone doesn't move, it is in HOLD mode. I hope it's still okay.

zeroos commented:
@MaEtUgR I've figured out the problem from yesterday: I did not realized, but some parameters were reset on the drone I took for testing and obstacle avoidance and optical flow was on. I disabled both and now the logs look nicer, here some logs from today's throw:
https://logs.px4.io/plot_app?log=9c0a0e9e-c25d-4855-821f-c00a5207aada

zeroos commented:
I've rebased to master hoping it would get rid of the MacOS build not going through, but it didn't help; I see also the same problem in other recent PRs...

zeroos commented:
@MaEtUgR , is there anything else I can do to help push this forward?

MaEtUgR reviewed (approved):
> is there anything else I can do to help push this forward?

No, thank you. It's as far modularized as it gets with the current architecture. It shows a good example of what needs to be exposed to a mode to be able to implement such a feature.

Is there an easy way to test this in simulation? Such that we can make sure it doesn't break e.g. when it's moved to its own mode?

zeroos commented:
Nice! Thank you once more for refining the code, the discussions during the PX4 summit, and for getting this through! I appreciate your work on the project. :)

To answer your question about testing: what I usually do in Gazebo is to turn on the throw parameter, arm the drone, and then remove the floor. The drone should fall slightly and then stabilize. It's not the same as throwing it, but good enough for a basic test.

In gazebo-classic it was possible to apply some instant force to an object. When I started developing the throw launch, I've used this feature. Unfortunately, some tweaking was required, because too much force resulted in the estimators not working well.

hamishwillee commented:
@zeroos @MaEtUgR Some questions for the docs:
1. When do you plan to submit them (you may have already and I have just missed the PR.
2. To keep in top of release changes in advance, I'm asking people to add release not for user-facing changes in the docs when they add information. That would go in here https://docs.px4.io/main/en/releases/main.html#simulation and match the pattern of the other releases.
3. Is this "solid". I.e. would I consider it experimental, or well tested, or what?

hamishwillee commented:
Hi @zeroos 
That helps a lot thanks - I work Weds/Thurs (today and tomorrow) each week on PX4 docs, so if you don't get a review on your docs until next week then it's nothing personal. Tag Matthias for technical review.

W.r.t. where should it go - yes, into multirotor section. In the docs try capture in a note what "throw" means - ie. what too low, to much roll mean. If these are parametized then what that means. List any parameters that you use that are significant. The flight mode docs are kind of a good example of what I mean here.

W.r.t experimental this is not well defined. But anything new usually needs some shakedown. So I propose we put a banner up the top stating "introduced after PX4 v1.14" and note that while this feature has been tested it has not had the broad testing that comes from having been in a few releases.

What would be particularly excellent would be to add tests in CI for this. Is that possible?

MaEtUgR commented:
> add tests in CI for this. Is that possible?

That would be ideal yes. That's why I was asking about the simulated throw which would need to be automated e.g. removing the floor by script.

zeroos commented:
The documentation needed some small refinements, but should be good now! The PR: https://github.com/PX4/PX4-user_guide/pull/2897

Having this added to the CI would be great. As the feature is now merged, I see it as the next step and I'll try to contribute it when the time allows.

dengrenjie321 commented:
Hi, I have a question.
the Fourth Way school https://docs.px4.io/main/en/flight_modes_mc/throw_launch.html#throw-launch-multicopter 。The test results are a failure, there seems to be no attitude control。
ulog:https://logs.px4.io/plot_app?log=707ce6aa-3269-429b-b3a5-61920157fa03

zeroos commented:
Hello!

I'm happy to see someone testing the feature! 

From the logs it seems that you were trying to throw it in the "hold/loiter" mode. I've had a mixed experience with that for two reasons:

1. It's harder for the drone to stabilize as in this mode it regulates both the position and the orientation. The drone that I was using was tuned in such a way that regulating the yaw significantly decreased thrust and it was harder for the drone to stabilize. Of course, this could be tuned, but it should be easier to have a successful throw in another mode, I usually use position mode.
2. The loitering point is set when you switch to the hold mode. If you do that before throwing the drone, the point will most probably be where you are. This results in a drone that behaves like a boomerang: you throw it away and then it tries to come back to you. The behaviour in "position" mode is closer to what you expect: the drone will stabilize more or less in the point you throw it to and will not try to come back.

dengrenjie321 commented:
Thank you very much for your reply.
I launched it in position mode, but it seems to have switched to loiter mode at the moment of launch. I need to find out the reason. Have you ever encountered this kind of issue。

zeroos commented:
Happens all the time! :) Good luck with the tests in the future!

dengrenjie321 commented:
Hi!
Today, I conducted a test, but unfortunately, it failed again. When the motors started, the aircraft did not adjust itself, and it crashed with the nose up. When I checked the logs, it seemed like there was an issue with altitude control. I'm not sure if it's because I chose the barometer or if I threw the drone too low. I placed the drone flat on the ground first, then lifted it after entering position mode, and threw it in a trajectory similar to a parabola. I'm unsure if this launching process is correct.
 I'm eagerly awaiting your response!!!
ulog:https://logs.px4.io/plot_app?log=a76cb897-0a4c-4fde-8581-41eb92be649d

zeroos commented:
It seems that the drone was thrown around 3m up and while going up already did around two flips. 

In my experience, and with the drones which we were using (PX4 Vision and simial X250 custom-built, overpowered drones), the drone can recover within 3m of height difference only when there's not much rotation when the motors start. With high rotation, you might need much more than 3m (but I would rather go for reducing the rotation while throwing than increasing the height :) ).

Unfortunately, the feature is not optimized for quick recovery, but it should still manage with the standard PX4 controller to work reasonably well.

From what I understand, the warning that you see at 2:37 only means that roll was too high, but the drone still tries to correct it.
